apollo-router/src/plugins/traffic_shaping/rate/service.rs

```diff
@@ -0,0 +1,69 @@
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::sync::RwLock;
+use std::task::Context;
+use std::task::Poll;
+
+use futures::ready;
+use tokio::time::Instant;
+use tower::Service;
+
+use super::future::ResponseFuture;
+use super::Rate;
+use crate::plugins::traffic_shaping::rate::error::RateLimited;
+
+#[derive(Debug)]
+pub(crate) struct RateLimit<T> {
+    pub(crate) inner: T,
+    pub(crate) rate: Rate,
+    pub(crate) window_start: Arc<RwLock<Instant>>,
+    pub(crate) previous_nb_requests: Arc<AtomicUsize>,
+    pub(crate) current_nb_requests: Arc<AtomicUsize>,
+}
+
+impl<S, Request> Service<Request> for RateLimit<S>
+where
+    S: Service<Request>,
+    S::Error: Into<tower::BoxError>,
+{
+    type Response = S::Response;
+    type Error = tower::BoxError;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let mut window_start = self.window_start.read().unwrap().elapsed();
+        let time_unit = self.rate.per();
+
+        if window_start > time_unit {
+            let new_window_start = Instant::now();
+            *self.window_start.write().unwrap() = new_window_start;
```

Comment:
it *might* be possible to avoid the lock by:
- storing a `AtomicU64` instead of an `Instant`. That can be done by noticing that we do not need the precision down to the nanosecond (we care at the millisecond level), nor up to the day. So instead of 2 u64, we can have it as one atomic int
- we can still get a duration by comparing those
- doing a "compare and swap" on it